README.md
# SydneyHUDFix
Fixed version of SydneyHUD after former team stopped updating this HUD

* Dom: Main Developer and Bugfixer after version 4.0.3

<h1 align="center">SydneyHUD</h1>
<p align="center">All-In-One PAYDAY2 HUD mod</p>

<p align="center">
<img src="https://img.shields.io/badge/BLT2-Compatibled-blue.svg?style=flat-square">
<a href="https://travis-ci.org/SydneyMOD/SydneyHUD" target="_blank"><img src="https://img.shields.io/travis/SydneyMOD/SydneyHUD.svg?style=flat-square"></a>
<a href="https://styleci.io/repos/75427456" target="_blank"><img src="https://styleci.io/repos/75427456/shield?branch=master"></a>
<img src="https://reposs.herokuapp.com/?path=SydneyMOD/SydneyHUD&style=flat-square&color=lightgray">
<a href="https://github.com/SydneyMOD/SydneyHUD/releases/latest" target="_blank"><img src="https://img.shields.io/github/release/SydneyMOD/SydneyHUD.svg?style=flat-square"></a>
<a href="https://github.com/SydneyMOD/SydneyHUD/blob/master/LICENSE" target="_blank"><img src="https://img.shields.io/github/license/SydneyMOD/SydneyHUD.svg?style=flat-square"></a>
</p>

### Features
* 360° view driving
* Always Show MOD Icons
* Ammo total left count without including clip
* Anti accidentally throwing grenades in stealth
* Anti bobble
* Armor/Inspire timers
* Auto enable Sentry Gun AP
* Auto laser on
* Changeable HUD scale
* Changeable Waypoint Text Color
* Chat feed system (Assault Phase / ECM battery / Inspire ACE CT / Down Warning)
* Chat Info
* ChatTime
* Contract broker quick link
* Corpse Remover Plus
* Custom Menu
* Damage Pop-up
* Down Counter / Detection risk indicator
* Down Warning
* Enhanced Assault Banner
* First Aid Kit notifyer
* Flashlight Extender
* Force enable / disable showing of weapon skins in lobby
* HPS Meters
* HUD Panel Lists
* HUD Tab Stats
* Health bar
* Hold to pick
* Ingame-voice indicators
* Improved assault indicator
* Interact Time Hint
* Kill-Counters
* Laser options
* Multi language Support (English / Japanese / French / Russian)
* Numerical Suspicion
* Numerical interaction countdown
* Pacified Civilian Markers
* Push to Interact
* Push to Interact Delay
* Reload interaction circle
* Screen Skipping
* Some graphical tweaks
* Spot Civilians
* Stamina Meter

### Feature Requests / Bug Reports
Please create issues [here](https://github.com/dominikvyhnak/SydneyHUDFix/issues) or Submit at [Steam Group](https://steamcommunity.com/groups/SydneyMOD/discussions/2/152390648090747618/)

### Downloads
Join SydneyMOD [Steam Group](https://steamcommunity.com/groups/SydneyMOD) and read about the group.

### Get in touch
* Steam Group [🔗](https://steamcommunity.com/groups/SydneyMOD)
* Discord [🔗](https://discord.gg/vF5a7gw)
* Dom's Discord [🔗](https://discord.gg/8CryHkM)
* Twitter [🔗](https://twitter.com/SydneyMOD)

### Contributors
# Current
* Dom: Main Developer & English Translator
* PoorPockets McNewHold: French Translator & Community Manager
* Sylvanas Windrunner: Russian Translator

# Former
* katabame: Main Developer
* BangL: Main Developer
* kash: Japanese Translator & Developer
* IPowis: English Translator & Developer
* hexco321: Feature contributer
* STaJIKeR1: Russian Translator
* exp1os1on: Russian Translator
* Kazenin: Portuguese Translator
* Nox-404: Bug fixer

### Special Thanks
SydneyHUD is using the following mods, codes & textures.
So, Big credit to these guys!

* 🐔**JackHUD** by BangL [🔗](https://steamcommunity.com/groups/jackhud)
* 💎**Hold to pick** by Malo [🔗](http://modworkshop.net/mydownloads.php?action=view_down&did=13518)
* 🏃**Anti-Bobble** by Schmuddel [🔗](http://modworkshop.net/mydownloads.php?action=view_down&did=15363)
* 🐺**WolfHUD** by KAMIKAZE94 [🔗](http://paydaymods.com/mods/298/wolfhud)
* 🚙**Driver's Neck Brace Removed** by EdisLeado [🔗](http://modworkshop.net/mydownloads.php?action=view_down&did=13569)
* 🔧**Always Show MOD Icons** by Slippy [🔗](http://modworkshop.net/mydownloads.php?action=view_down&did=13975)
* 👦🏻**Civilian Marking in Stealth** by EdisLeado [🔗](http://modworkshop.net/mydownloads.php?action=view_down&did=13517)
* 🚑**Replenished** by FishTaco [🔗](http://modworkshop.net/mydownloads.php?action=view_down&did=15309)
* 💀**Swan Song Effect** by Simon [🔗](http://modworkshop.net/mydownloads.php?action=view_down&did=681)
* 📦**No Bag Bendy Tilt** by mluzarow [🔗](https://github.com/mluzarow/No-Bag-Bendy-Tilt)
* 💉**FAK-Notifier** by SlashNephy [🔗](https://github.com/SlashNephy/FAK-Notifier)
* 💬**Chat Info** by Nebby [🔗](https://modworkshop.net/mydownloads.php?action=view_down&did=19181)
* 🔺**Better Assault Indicator** by Dom [🔗](https://modworkshop.net/mydownloads.php?action=view_down&did=22712)
* Menu code from **VoidUI** by VoidUI Team [🔗](https://modworkshop.net/mod/20997)
* many of inspired mod

**AND YOU.**

changelog.txt
4.2.8 (xx. 06. 2022):
Added "Reduce Shotgun Spam" optimization

4.2.7 (11. 02. 2022):
Updated BAI code to r134
Removed Down Counter
Fixed an error string in "Fixing PocoHUD settings conflict" screen
Fixed 'mods/SydneyHUD/lua/HUDTeammate.lua:600: attempt to index field '_detection_counter' (a nil value)' crash
Fixed a typo in "5sec remining" reminder
Fixed a UI bug with weapon reload interaction when the weapon reload was interrupted

4.2.6 (04. 04. 2021):
Added name for the Triad Thugs
Fixed an issue where kills are counted as double kills
Dragon Heist: Fixed an issue where C4 Timer is stucked in HUDList when picked up

4.2.5 (18. 12. 2020):
Added Spanish translation
Added trackers for the enemies in Buluc's Mansion
Possible fix for crash "set_target_alpha()"
Fixed an issue where teammate's downs were not counted
Fixed an issue where Russian enemies did not have proper name
Fixed an issue where "Hide Hostage Panel" option did not work
Menu: Fixed an issue where a description text was not used
Menu: Fixed an issue an error string in the assault panel options 

4.2.4 (14. 08. 2020):
Added Kingpin Effect
Fixed an issue where Quick Fix buff did nothing in HUDList
Fixed Swan Song flashing texture
Removed "Buy all assets"
Removed load and save preplanning keybinds

4.2.3 (09. 07. 2020):
Added trackers for Romanov Armor in Border Crossing heist and Federal Medals in Breakfast in Tijuana heist
Improved options changing while you are in heíst
Possible fix for a crash that could happen at the end of Holdout
Menu: Fixed an issue where Color Selection for Enemy Snipers is cut off

4.2.2 (19. 05. 2020):
Added trackers for the new loot added in San Martín Bank Heist
Added HUDList Minion Color
Merged HUDList's Civilian Color and Enemy Color into one menu
Fixed an issue where interaction color is not saved properly
Fixed an issue where options related to "Suspicion" are not working properly
Possible fix for 'mods/SydneyHUD/lua/lib/managers/hud/HUDAssaultCorner.lua:801: attempt to index field 'hud' (a nil value)'

4.2.1 (08. 05. 2020):
Fixed an issue where some text got mixed up in Custom Turret Color
Fixed an issue "Ammo Bags" option in HUDList is not loading and saving correctly
Fixed a crash when you changed "Ammo Bags" option in HUDList
Fixed a crash when you changed options in Infiltrator in HUDList

4.2.0 (08. 05. 2020):
Added "Kill Counter Color" option
Redesigned mod options
Removed "Experimental Tweaks"
Fixed a crash when you were client

4.1.9 (30. 04. 2020):
Added Assault Panel color animation
"Time Left" is now accurate in Holdout (host only)
Time Left on client is more accurate in Safe House Raid and Alaskan Deal heists
Fixed a crash that could happen when you started Holdout mission

4.1.8 (18. 04. 2020):
Fixed an issue where your weapon laser is always green
Fixed an issue where your weapon laser could sometimes show up as white

4.1.7 (11. 04. 2020):
Re-added back rainbow color laser for players
Fixed an issue where "Block Blood Decals" option was not saved correctly
Fixed a crash when you were playing with KineticHUD host and had "Enhanced Assault Banner" option on

4.1.6 (22. 03. 2020):
Added trackers for the units introduced in Cartel Business Update
Added support for Federales faction
Improved faction detection
Fixed crashing as client when you returned to lobby
Fixed a crash when you reset mod options to default values
Fixed an issue where reload timer did not hide when you interrupted reload of a shell shotgun

4.1.5 (31. 12. 2019):
Added an animation for left list in HUDList
Redesigned settings menu a bit
Removed support for normal BLT
Fixed crashes related to interactions
Fixed an issue where time is visible in PONR banner

4.1.4 (21. 11. 2019):
Added Optimization Tweaks
Added text warning about HUDList options working differently
Updated Assault Banner code to BAI73
Optimized "Show Reload Time As Interaction" option
Optimized "Show Melee Charging Time As Interaction" option
Fixed an issue where the interaction bar was stuck on the screen when a Cloaker kicked you
Fixed an issue where the interaction sound did not stop when you pressed F key or G key
Fixed an text issue where Medic Dozer and Minigun Dozer did not have "Murkywater" prefix in the health bar when playing Murkywater heists
Fixed an issue on Border Crossing and Border Crystals where bolivian guards were not counted as guards with pagers in HUDList

4.1.3 (08. 11. 2019):
Added missing name for the new enemy introduced in the new heists
Enemies in the Border Crossing and Border Crystals now have correct names
Fixed an issue where hostage panel overlaps with HUDList
Fixed "mods/SydneyHUD/lua/lib/managers/hud/HUDAssaultCorner.lua:791: attempt to call method 'SetCompatibleHost' (a nil value)" crash

4.1.2 (06. 11. 2019):
All color values are RGB-compliant
Added "Shuffle Music" option and keybind
Added a warning message when the saved data is not compatible
Updated Assault Banner code to BAI68
Fixed captain's buff icon alignment when the assault panel is centered
Fixed an issue where hostage panel is not visible when HUDList is disabled

HUDList:
Fixed an issue where multi-choice options were not correctly parsed
Fixed an issue where OVERDOG skill option did not change in HUDLIst
Fixed an issue where "Show Hostages" option was a toggle instead of multi-choice
Fixed an issue where changing "Show Loot" option changed "Show Body Bags" optionFixed an issue where Swan Song Effect is not visible when HUDList is disabled
Fixed Buff List Scale and Right List Scale options not saving properly

4.1.1 (21. 08. 2019):
Added Music Tweaks
Updated Assault Banner code to BAI r67
Optimized Kill counter
Fixed "Press to Hold" not appearing
Removed duplicate timer waypoints on certain heists
Fixes to interaction text
Fixed an issue where Sentry Guns don't automatically change ammo to AP rounds
Fixed an issue where enemy health bar is visible when you are in custody
Fixed crashes related to HUDTeammate

4.1.0 (23. 02. 2019):

lua/HUDManager.lua
function HUDManager:add_waypoint(id, data)
    if self._hud.waypoints[id] then
        self:remove_waypoint(id)
    end
    local hud = managers.hud:script(PlayerBase.PLAYER_INFO_HUD_PD2)
    if not hud then
        self._hud.stored_waypoints[id] = data
        return
    end
    local waypoint_panel = hud.panel
    local icon = data.icon or "wp_standard"
    local text = ""
    local texture_rect
    icon, texture_rect = tweak_data.hud_icons:get_icon_data(icon, {
        0,
        0,
        32,
        32
    })
    local bitmap = waypoint_panel:bitmap({
        name = "bitmap" .. id,
        texture = icon,
        texture_rect = texture_rect,
        layer = 0,
        w = texture_rect[3],
        h = texture_rect[4],
        blend_mode = data.blend_mode,
        rotation = 360
    })
    local arrow_icon, arrow_texture_rect = tweak_data.hud_icons:get_icon_data("wp_arrow")
    local arrow = waypoint_panel:bitmap({
        name = "arrow" .. id,
        texture = arrow_icon,
        texture_rect = arrow_texture_rect,
        color = data.color or Color.white:with_alpha(0.75),
        visible = false,
        layer = 0,
        w = arrow_texture_rect[3],
        h = arrow_texture_rect[4],
        blend_mode = data.blend_mode,
        rotation = 360
    })
    local distance
    local r = SydneyHUD:GetOption("waypoint_color_r")
    local g = SydneyHUD:GetOption("waypoint_color_g")
    local b = SydneyHUD:GetOption("waypoint_color_b")
    if data.distance then
        distance = waypoint_panel:text({
            name = "distance" .. id,
            text = "16.5",
            color = data.color or Color(r, g, b), --Color(1, 1, 0.65882355, 0),
            font = tweak_data.hud.medium_font_noshadow,
            font_size = tweak_data.hud.default_font_size,
            align = "center",
            vertical = "center",
            w = 128,
            h = 24,
            layer = 0,
            blend_mode = data.blend_mode,
            rotation = 360
        })
        distance:set_visible(false)
    end
    local timer = data.timer and waypoint_panel:text({
        name = "timer" .. id,
        text = (math.round(data.timer) < 10 and "0" or "") .. math.round(data.timer),
        font = tweak_data.hud.medium_font_noshadow,
        font_size = 32,
        align = "center",
        vertical = "center",
        w = 32,
        h = 32,
        layer = 0,
        rotation = 360
    })
    text = waypoint_panel:text({
        name = "text" .. id,
        text = utf8.to_upper(" " .. text),
        font = tweak_data.hud.small_font,
        font_size = tweak_data.hud.small_font_size,
        align = "center",
        vertical = "center",
        w = 512,
        h = 24,
        layer = 0,
        rotation = 360
    })
    local _, _, w, _ = text:text_rect()
    text:set_w(w)
    local h
    w, h = bitmap:size()
    self._hud.waypoints[id] = {
        init_data = data,
        state = data.state or "present",
        present_timer = data.present_timer or 2,
        bitmap = bitmap,
        arrow = arrow,
        size = Vector3(w, h, 0),
        text = text,
        distance = distance,
        timer_gui = timer,
        timer = data.timer,
        pause_timer = not data.pause_timer and data.timer and 0,
        position = data.position,
        unit = data.unit,
        no_sync = data.no_sync,
        move_speed = 1,
        radius = data.radius or 160
    }
    self._hud.waypoints[id].init_data.position = data.position or data.unit:position()
    local t = {}
    for _, fdata in pairs(self._hud.waypoints) do
        if fdata.slot then
            t[fdata.slot] = fdata.text:w()
        end
    end
    for i = 1, 10 do
        if not t[i] then
            self._hud.waypoints[id].slot = i
            break
        end
    end
    self._hud.waypoints[id].slot_x = 0
    if self._hud.waypoints[id].slot == 2 then
        self._hud.waypoints[id].slot_x = t[1] / 2 + self._hud.waypoints[id].text:w() / 2 + 10
    elseif self._hud.waypoints[id].slot == 3 then
        self._hud.waypoints[id].slot_x = -t[1] / 2 - self._hud.waypoints[id].text:w() / 2 - 10
    elseif self._hud.waypoints[id].slot == 4 then
        self._hud.waypoints[id].slot_x = t[1] / 2 + t[2] + self._hud.waypoints[id].text:w() / 2 + 20
    elseif self._hud.waypoints[id].slot == 5 then
        self._hud.waypoints[id].slot_x = -t[1] / 2 - t[3] - self._hud.waypoints[id].text:w() / 2 - 20
    end
end

local CharacterData = {
    ["security"] = "SECURITY GUARD",
    ["security_undominatable"] = "SECURITY GUARD",
    ["gensec"] = "GENSEC UNIT",
    ["cop"] = "POLICE OFFICER",
    ["cop_scared"] = "POLICE OFFICER",
    ["cop_famale"] = "POLICE OFFICER",
    ["inside_man"] = "INSIDER",
    ["fbi"] = "FBI UNIT",
    ["swat"] = "SWAT UNIT",
    ["heavy_swat"] = "HEAVY SWAT",
    ["fbi_swat"] = "FBI SWAT",
    ["fbi_heavy_swat"] = "FBI HEAVY SWAT",
    ["city_swat"] = "CITY SWAT / MURKYWATER",
    ["sniper"] = "SNIPER",
    ["gangster"] = "GANGSTER",
    ["biker"] = "BIKER",
    ["biker_escape"] = "BIKER",
    ["mobster"] = "RUSSIAN MOBSTER",
    ["mobster_boss"] = "THE COMMISSAR",
    ["hector_boss"] = "HECTOR",
    ["hector_boss_no_armor"] = "HECTOR",
    ["tank"] = "BULLDOZER",
    ["tank_hw"] = "HEADLESS DOZER",
    ["tank_mini"] = "MINIGUN DOZER",
    ["tank_medic"] = "MEDIC DOZER",
    ["spooc"] = "CLOAKER",
    ["shield"] = "SHIELD",
    ["phalanx_minion"] = "WINTER'S UNIT",
    ["phalanx_vip"] = "CAPTAIN WINTERS",
    ["taser"] = "TASER",
    ["civilian"] = "CIVILIAN",
    ["civilian_female"] = "CIVILIAN",
    ["melee_box"] = "CIVILIAN",
    ["robbers_safehouse"] = "CIVILIAN",
    ["boris"] = "BORIS",
    ["bank_manager"] = "BANK MANAGER",
    ["drunk_pilot"] = "DRUNK PILOT",
    ["escort"] = "ESCORT",
    ["escort_undercover"] = "TAXMAN",
    ["russian"] = "DALLAS",
    ["german"] = "WOLF",
    ["spanish"] = "CHAINS",
    ["american"] = "HOUSTON",
    ["jowi"] = "JOHN WICK",
    ["old_hoxton"] = "HOXTON",
    ["female_1"] = "CLOVER",
    ["dragan"] = "DRAGAN",
    ["jacket"] = "JACKET",
    ["bonnie"] = "BONNIE",
    ["sokol"] = "SOKOL",
    ["dragon"] = "JIRO",
    ["bodhi"] = "BODHI",
    ["jimmy"] = "JIMMY",
    ["sydney"] = "SYDNEY",
    ["wild"] = "RUST",
    ["chico"] = "SCARFACE",
    ["biker_boss"] = "BIKER BOSS",
    ["chavez_boss"] = "CHAVEZ",
    ["bolivian"] = "SOSA THUG",
    ["bolivian_indoors"] = "SOSA SECURITY",
    ["drug_lord_boss"] = "SOSA BOSS",
    ["drug_lord_boss_stealth"] = "SOSA BOSS",
    ["old_hoxton_mission"] = "", -- Fixed dynamically
    ["medic"] = "MEDIC",
    ["spa_vip"] = "CHARON",
    ["spa_vip_hurt"] = "CHARON",
    ["swat_turret"] = "SWAT TURRET"
}

local function FixNames()
    local level_id, difficulty = Global.game_settings.level_id, Global.game_settings.difficulty
    if SydneyHUD:IsOr(level_id, "hox_1", "hox_2") then
        CharacterData["old_hoxton_mission"] = "HOXTON"
    end
    if SydneyHUD:IsOr(difficulty, "easy_wish", "overkill_290") then -- Mayhem or Death Wish
        CharacterData["city_swat"] = "GENSEC SWAT UNIT"
        CharacterData["fbi_heavy_swat"] = "GENSEC HEAVY SWAT"
    end
    if difficulty == "sm_wish" then -- Death Sentence
        CharacterData["swat"] = "ZEAL SWAT UNIT"
        CharacterData["heavy_swat"] = "ZEAL HEAVY SWAT"
        CharacterData["tank"] = "ZEALDOZER"
        CharacterData["spooc"] = "ZEAL CLOAKER"
        CharacterData["shield"] = "ZEAL SHIELD"
        CharacterData["taser"] = "ZEAL TASER"
    end
    if SydneyHUD:IsOr(level_id, "kosugi", "pbr", "dark", "des", "bph", "vit") then -- Shadow Raid, Beneath the Mountain, Murky Station, Henry's Rock, Hell's Island, The White House
        CharacterData["fbi"] = "MURKYWATER LIGHT"
        CharacterData["fbi_swat"] = "MURKYWATER LIGHT"
        CharacterData["city_swat"] = "MURKYWATER"
        CharacterData["fbi_heavy_swat"] = "MURKYWATER HEAVY"
        CharacterData["medic"] = "MURKYWATER MEDIC"
        CharacterData["taser"] = "MURKYWATER TASER"
        CharacterData["tank"] = "MURKYWATER BULLDOZER"
        CharacterData["spooc"] = "MURKYWATER CLOAKER"
        CharacterData["shield"] = "MURKYWATER SHIELD"
        CharacterData["swat"] = "MURKYWATER LIGHT" -- Death Sentence
        CharacterData["heavy_swat"] = "MURKYWATER HEAVY" -- Death Sentence
        if level_id == "pbr" then -- Beneath the Mountain
            CharacterData["old_hoxton_mission"] = "LOCKE"
        end
    end
    if level_id == "mad" then -- Boiling Point
        CharacterData["fbi_swat"] = "RUSSIAN LIGHT"
        CharacterData["city_swat"] = "RUSSIAN LIGHT"
        CharacterData["fbi_heavy_swat"] = "RUSSIAN HEAVY"
        CharacterData["tank"] = "RUSSIAN BULLDOZER"
        CharacterData["sniper"] = "RUSSIAN SNIPER"
        CharacterData["spooc"] = "RUSSIAN CLOAKER"
        CharacterData["shield"] = "RUSSIAN SHIELD"
        CharacterData["taser"] = "RUSSIAN TASER"
        CharacterData["medic"] = "RUSSIAN MEDIC"
        CharacterData["swat_turret"] = "RUSSIAN TURRET"
    end
end

FixNames()

Hooks:PostHook(HUDManager, "_player_hud_layout", "uHUDPostHUDManagerPlayerInfoHUDLayout", function(self)
    local unit_health_main = managers.hud:script(PlayerBase.PLAYER_INFO_HUD_PD2).panel:panel({
        name 	= "unit_health_main",
        halign 	= "grow",
        valign 	= "grow"
    })
    self._unit_health_panel = unit_health_main:panel({
        name 	= "unit_health_panel",
        visible = false
    })
    self._unit_health = self._unit_health_panel:bitmap({
        name 			= "unit_health",
        texture 		= "guis/textures/pd2/healthshield",
        texture_rect 	= {
                            2,
                            18,
                            232,
                            11
                        },
        blend_mode 		= "normal"
    })

    if SydneyHUD:GetOption("health_bar_color") == 1 then
        self._health_text_rect = { 2, 18, 232, 11 }
    elseif SydneyHUD:GetOption("health_bar_color") == 2 then
        self._health_text_rect = { 2, 34, 232, 11 }
    elseif SydneyHUD:GetOption("health_bar_color") == 3 then
        self._health_text_rect = { 2, 82, 232, 11 }
    elseif SydneyHUD:GetOption("health_bar_color") == 4 then
        self._health_text_rect = { 2, 98, 232, 11 }
    end

    self._unit_shield = self._unit_health_panel:bitmap({
        name 			= "unit_shield",
        texture 		= "guis/textures/pd2/healthshield",
        texture_rect 	= {
                            1,
                            1,
                            234,
                            13
                        },
        blend_mode 		= "normal"
    })
    self._unit_health_text = self._unit_health_panel:text({
        name 		= "unit_health_text",
        text 		= "N/A",
        blend_mode 	= "normal",
        alpha 		= 1,
        halign 		= "right",
        font 		= "fonts/font_medium_shadow_mf",
        font_size 	= 20,
        color 		= Color.white,
        align 		= "center",
        layer 		= 1
    })
    self._unit_health_enemy_text = self._unit_health_panel:text({
        name 		= "unit_health_enemy_text",
        text 		= "SWAT VAN TURRET",
        blend_mode 	= "normal",
        alpha 		= 1,
        halign 		= "left",
        font 		= "fonts/font_medium_shadow_mf",
        font_size 	= 22,
        color 		= Color.white,
        align 		= "center",
        layer 		= 1
    })
    self._unit_health_enemy_location = self._unit_health_panel:text({
        name 		= "unit_health_enemy_location",
        text 		= "^",
        blend_mode 	= "normal",
        visible 	= false,
        alpha 		= 0.75,
        halign 		= "center",
        font 		= "fonts/font_medium_shadow_mf",
        font_size 	= 20,
        color 		= Color.white,
        align 		= "center",
        layer 		= 1
    })
    local hx, hy, hw, hh = self._unit_health_text:text_rect()
    local ex, ey, ew, eh = self._unit_health_enemy_text:text_rect()
    local lx, ly, lw, lh = self._unit_health_enemy_location:text_rect()
    self._unit_health_text:set_size(hw, hh)
    self._unit_health_enemy_text:set_size(ew, eh)
    self._unit_health_enemy_location:set_size(lw, lh)
    self._unit_health:set_w(self._unit_health:w() - 2)
    self._unit_health:set_center(self._unit_health_panel:center_x(), self._unit_health_panel:center_y() - 190)
    self._unit_shield:set_center(self._unit_health_panel:center_x(), self._unit_health_panel:center_y() - 190)
    self._unit_health_text:set_right(self._unit_shield:right())
    self._unit_health_text:set_bottom(self._unit_shield:top())
    self._unit_health_enemy_text:set_left(self._unit_shield:left())
    self._unit_health_enemy_text:set_bottom(self._unit_shield:top())
    self._unit_health_enemy_location:set_center_x(self._unit_shield:center_x())
    self._unit_health_enemy_location:set_top(self._unit_shield:bottom())
end)

function HUDManager:set_unit_health_visible(visible)
    if SydneyHUD:GetOption("show_enemy_health") and visible and not self._unit_health_visible then
        self._unit_health_visible = true
        self._unit_health_panel:stop()
        self._unit_health_panel:animate(function(p)
            self._unit_health_panel:set_visible(true)
            over(0.25, function(o)
                self._unit_health_panel:set_alpha(math.lerp(self._unit_health_panel:alpha(), 1, o))
            end)
        end)
    elseif not visible and self._unit_health_visible then
        self._unit_health_visible = nil
        self._unit_health_panel:stop()
        self._unit_health_panel:animate(function(p)
            if self._unit_health_panel:alpha() >= 0.9 then
                over(0.5, function(o) end)
            end
            over(1.5, function(o)
                self._unit_health_panel:set_alpha(math.lerp(self._unit_health_panel:alpha(), 0, o))
            end)
            self._unit_health_panel:set_visible(false)
        end)
    end
end


function HUDManager:set_unit_health(current, total, tweak_table)
    if not current or not total then
        return
    end
    local enemy = CharacterData[tweak_table] or tweak_table
    local _r = current / total
    local r = self._unit_health:width()
    local rn = (self._health_text_rect[3] - 2) * _r
    self._unit_health_enemy_text:set_text(enemy)
    self._unit_health_text:set_text(string.format("%d / %d (%d%%)", current * 10, total * 10, current/total * 100))
    local hx, hy, hw, hh = self._unit_health_text:text_rect()
    local ex, ey, ew, eh = self._unit_health_enemy_text:text_rect()
    self._unit_health_text:set_size(hw, hh)
    self._unit_health_enemy_text:set_size(ew, eh)
    self._unit_health_text:set_right(self._unit_shield:right())
    self._unit_health_text:set_bottom(self._unit_shield:top())
    self._unit_health_enemy_text:set_left(self._unit_shield:left())
    self._unit_health_enemy_text:set_bottom(self._unit_shield:top())
    self._unit_health_text:set_color(current == 0 and Color.red or Color.white)
    self._unit_health:stop()

    if rn > r then
        self._unit_health:animate(function(p)
            over(0.5, function(o)
                self._unit_health:set_w(math.lerp(r, rn, o))
                self._unit_health:set_texture_rect(self._health_text_rect[1], self._health_text_rect[2], math.lerp(r, rn, o), self._health_text_rect[4])
            end)
        end)
    end

    self._unit_health:set_w(_r * (self._health_text_rect[3] - 2))
    self._unit_health:set_texture_rect(self._health_text_rect[1], self._health_text_rect[2], self._health_text_rect[3] * _r, self._health_text_rect[4])
end

function HUDManager:set_unit_health_rotation(angle)
    self._unit_health_enemy_location:set_rotation(angle)
end